packages/examples/cvat/recording-oracle/src/handlers/process_intermediate_results.py

```diff
@@ -400,20 +392,33 @@ def process_intermediate_results(  # noqa: PLR0912
     if updated_gt_stats and updated_gt_stats != initial_gt_stats:
         cvat_task_id_to_failed_val_frames = {}  # cvat_task_id: [val_frame_id, ...]
         for (cvat_task_id, val_frame_id), value in updated_gt_stats.items():
-            if value["failed_attempts"] >= Config.validation.gt_ban_threshold:
+            if value >= Config.validation.gt_ban_threshold:
                 cvat_task_id_to_failed_val_frames.setdefault(cvat_task_id, []).append(val_frame_id)
 
         for cvat_task_id, val_frame_ids in cvat_task_id_to_failed_val_frames.items():
-            cvat_api.disable_validation_frames(cvat_task_id, frames_to_disable=val_frame_ids)
-            jobs_with_honeypots_to_be_shuffled = []
-
-            for val_frame_id in val_frame_ids:
-                jobs_with_honeypots_to_be_shuffled.extend(
-                    validation_result.val_frame_id_to_jobs[val_frame_id]
-                )
+            task_validation_layout = validation_result.task_id_to_val_layout[cvat_task_id]
+            intersection = set(val_frame_ids) & set(task_validation_layout.disabled_frames)
+            if intersection:
+                logger.error(f"Unexpected case: frames {intersection} were disabled earlier")
+
+            updated_disable_frames = task_validation_layout.disabled_frames + val_frame_ids
+            not_disabled_frames = list(
+                set(task_validation_layout.validation_frames) - set(updated_disable_frames)
+            )
 
-            jobs_with_honeypots_to_be_shuffled = set(jobs_with_honeypots_to_be_shuffled)
-            cvat_api.shuffle_honeypots_in_jobs(jobs_with_honeypots_to_be_shuffled)
+            rng = np.random.default_rng()
+            upd_honeypot_real_frames = [
+                frame
+                if frame not in updated_disable_frames
+                else int(rng.choice(not_disabled_frames))
+                for frame in task_validation_layout.honeypot_real_frames
```

Comment:
It's possible to get several identical honeypots in a job with such an algorithm. Consider changing all the task frames (e.g. using the random method) or excluding frames that are already used in a job. For the latter you'll need to remember segment size somewhere, as in CVAT it will be 0.

Reply:
I just realized that using "random" method will update all the jobs, including already finished ones. This can be a problem if we export these jobs after they are updated, as their honeypot frames will have no annotations.

Reply:
@zhiltsov-max, I see several solutions for this problem:
1. Utilize GET `/api/tasks/<id>/data/meta` endpoint to obtain metadata for the whole task and use `segment_size` from the task details. In that case, we need to update CVAT logic to return the actual segment size (annotation frames + honeypots). It's not clear to me why we manually set the value to 0.
2. If we cannot return the actual `segment_size` from CVAT, we can define it on the Exchange Oracle side (in generall it will be `manifest.annotation.job_size` + `manifest.validation.val_size`) and then pass it to the Recording Oracle via the `annotation_meta` file. I don't think this is the desired solution because this logic should be encapsulated inside CVAT.
3. Revert the previous approach and obtain a validation layout for each job. This approach also looks undesired since it requires +N requests to CVAT that can be throttled.
4. Introduce a parameter in PATCH `/api/tasks/<id>/validation_layout` that will define whether jobs with the acceptance stage should be included in the job queryset in which honeypots should be shuffled.  

Which approach do you prefer?

Reply:
I also see a small problem we shuffle honeypots in jobs before resetting the assignee. That means that an assignee will be able to open a job on CVAT and define which frames were honeypots/see new ones (before Exchange oracle handles webhook from Recording oracle).

Reply:
> It's not clear to me why we manually set the value to 0

There are 2 reasons for this:
- the value would be different from what's requested by the user during task creation (e.g. 4 normal + 2 validation frames would become 6)
- there are not segments in the task, really. The frames come in the random order, they are not slices of some range. This indicates that the "segments" are special

I think the second approach is the best for now. 4 is also somewhat reasonable, but the suggested design looks clumsy to me. Maybe we could change it to a list of relevant job ids instead, but it has some scalability issues.

Reply:
> I also see a small problem we shuffle honeypots in jobs before resetting the assignee. That means that an assignee will be able to open a job on CVAT and define which frames were honeypots/see new ones (before Exchange oracle handles webhook from Recording oracle).

Not relevant anymore